singer_sdk/streams/core.py

```diff
@@ -1161,6 +1265,33 @@ def get_records(self, context: dict | None) -> Iterable[dict | tuple[dict, dict]
         """
         pass
 
+    def get_batches(
+        self,
+        context: dict | None = None,
+    ) -> Iterable[tuple[BaseBatchFileEncoding, list[str]]]:
+        """Batch generator function.
+
+        Developers are encouraged to override this method to customize batching
+        behavior for databases, bulk APIs, etc.
+
+        Args:
+            context: Stream partition or context dictionary.
+
+        Yields:
+            A tuple of (encoding, manifest) for each batch.
+        """
+        for chunk in lazy_chunked_generator(
+            self._sync_records(context, write_messages=False),
+            self.batch_size,
+        ):
+            filename = f"output/{self.name}-{uuid4()}.json.gz"
+
+            # TODO: Determine compression from config.
+            with gzip.open(filename, "wb") as f:
+                f.writelines((json.dumps(record) + "\n").encode() for record in chunk)
+
+            yield JSONLinesEncoding(compression="gzip"), [filename]
```

Comment:
I'm adding a single file in each manifest.

See https://github.com/meltano/sdk/discussions/859#discussioncomment-3512700.

Reply:
I replied there in the discussion. Thank you _very_ much for logging that because I think that discussion is extremely valuable.

My take on that is threefold:

1. Yes, totally okay by me if the default SDK sends one file per batch message. (List-of-one would be my recommendation.)
2. Totally fine if it this is not configurable by the developer. (Max-records-per-batch and/or max-mb-per-batch could be future config options.)
3. I want to still keep the door open for taps that emit multiple files at once to still support multiple files per batch message.

Reply:
I replied there in the discussion. Thank you _very_ much for logging that because I think that discussion is extremely valuable.

My take on that is threefold:

1. Yes, totally okay by me if the default SDK sends one file per batch message. (List-of-one would be my recommendation.)
2. Totally fine if it this is not configurable by the developer. (Max-records-per-batch and/or max-mb-per-batch could be future config options.)
3. I want to still keep the door open for taps that emit multiple files at once to still support multiple files per batch message.